File: c/pseudoPalindromicPaths.cpp
// https://leetcode.com/problems/pseudo-palindromic-paths-in-a-binary-tree/
// 2024/01/24

#include <iostream>
#include <unordered_map>

#include "TreeNode.hpp"
using namespace std;

class Solution {
    int result;

   public:
    int pseudoPalindromicPaths(TreeNode* root) {
        result = 0;
        dfs(root, 0);
        root->left = root->right = 0;
        return result;
    }

    bool isPowerOfTwo(int n) { return n == 0 ? true : !(n & (n - 1)); }

    void dfs(TreeNode* root, int value) {
        if (root == nullptr) return;

        value ^= (1 << root->val);

        if (root->left == nullptr && root->right == nullptr) {
            result += static_cast<int>(isPowerOfTwo(value));
            return;
        }

        dfs(root->left, value);
        dfs(root->right, value);
    }
};

int main() {
    Solution solution;
    TreeNode* root =
        new TreeNode(2, new TreeNode(3, new TreeNode(3), new TreeNode(1)),
                     new TreeNode(1, nullptr, new TreeNode(1)));
    printTree(root);
    int result = solution.pseudoPalindromicPaths(root);
    cout << result << endl;
    return 0;
}

File: rust/src/bin/pseudo_palindromic_paths.rs
// https://leetcode.com/problems/pseudo-palindromic-paths-in-a-binary-tree/
// 2024/01/24

mod tree_node;
use crate::tree_node::TreeNode;

use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn pseudo_palindromic_paths(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut stack: Vec<(Rc<RefCell<TreeNode>>, usize)> = vec![];
        if let Some(node) = root {
            stack.push((node, 0));
        }

        let mut count = 0;

        while let Some((node, freq)) = stack.pop() {
            let mut node_ref = node.borrow_mut();
            let freq = freq ^ (1 << node_ref.val);

            if node_ref.left.is_none() && node_ref.right.is_none() {
                count += ((freq & (freq - 1)) == 0) as i32;
                continue;
            }

            if let Some(right) = node_ref.right.take() {
                stack.push((right, freq));
            }
            if let Some(left) = node_ref.left.take() {
                stack.push((left, freq));
            }
        }

        return count;
    }
}

pub struct Solution {}

fn main() {
    println!(
        "{:?}",
        Solution::pseudo_palindromic_paths(TreeNode::from_array(
            &[Some(2), Some(3), Some(1), Some(3), Some(1), None, Some(1)],
            0,
        ))
    ); // 2

    println!(
        "{:?}",
        Solution::pseudo_palindromic_paths(TreeNode::from_array(
            &[
                Some(2),
                Some(1),
                Some(1),
                Some(1),
                Some(3),
                None,
                None,
                None,
                None,
                None,
                Some(1),
            ],
            0,
        ))
    ); // 1
}
